keps/sig-auth/267-kubelet-server-certificate-bootstrap-rotation/README.md

```diff
@@ -0,0 +1,292 @@
+
+This proposal covers a process for generating a key locally and then issuing a
+Certificate Signing Request to the cluster API server to get an associated
+certificate signed by the cluster Certificate Authority. Also, as certificates
+approach expiration, the same mechanism will be used to request an updated
+certificate.
+
+### Goals
+
+- Allow Nodes/Kubelet to request a certificate to an external authority instead of using self signed certificates during bootstrap.
+- Allow Nodes/Kubelet to request renewals of their certificates to an external authority.
+
+
+## Design Details
+
+Kubernetes v1.8 and higher kubelet implements features for enabling rotation of its client and/or serving certificates.
+
+You can configure the kubelet to rotate its client certificates by creating new CSRs as its existing credentials expire. To enable this feature, user can enable the rotateCertificates field of kubelet configuration file.
+
+The CSR approving controllers implemented in core Kubernetes do not approve node serving certificates for security reasons. To use RotateKubeletServerCertificate operators need to run a custom approving controller, or manually approve the serving certificate requests.
+
+A deployment-specific approval process for kubelet serving certificates should typically only approve CSRs which:
+
+1. are requested by nodes (ensure the spec.username field is of the form system:node:<nodeName> and spec.groups contains system:nodes)
+
+2. request usages for a serving certificate (ensure spec.usages contains server auth, optionally contains digital signature and key encipherment, and contains no other usages)
+
+3. only have IP and DNS subjectAltNames that belong to the requesting node, and have no URI and Email subjectAltNames (parse the x509 Certificate Signing Request in spec.request to verify subjectAltNames)
+
+### Test Plan
+
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+
+##### Unit tests
+
+- `k8s.io/kubernetes/pkg/kubelet/certificate/kubelet.go`: `2024-09-12` - `57.7`
+- k8s.io/kubernetes/pkg/kubelet/certificate/transport.go: `2024-09-12` - `64.1`
+- k8s.io/kubernetes/pkg/kubelet/certificate/bootstrap/bootstrap.go: `2024-09-12` - `50`
+
+- `<package>`: `<date>` - `<test coverage>`
+
+##### Integration tests
+
+##### e2e tests
+
+Any cluster running e2e tests with the feature enabled will be exercising the feature.
+
+A job using the built-in CSR approver will be added exercising all the Conformance e2e tests.
+
+### Graduation Criteria
+
+#### Alpha
+
+- Feature implemented behind a feature flag
+- Initial e2e tests completed and enabled
+
+#### Beta
+
+- Real world usage
+
+#### GA
```

Comment:
corresponding metrics needs to be reviewed and GA-ed

Reply:
is this a new requirement?

We have most of the metrics in alpha
```
kubernetes$ grep -r StabilityLevel pkg/ | cut -d\: -f3 | sort | uniq -c
     11  compbasemetrics.ALPHA,
      2     metrics.ALPHA,
    179  metrics.ALPHA,
      1  metrics.BETA,
      1  metrics.INTERNAL,
      1     metrics.STABLE,
     14  metrics.STABLE,
```

Reply:
I [asked](https://github.com/kubernetes/kubernetes/pull/127146#discussion_r1768026291) Han about this the other day because I also wanted to know and he said that metrics can stay in alpha even when features graduate. There's no process that ensures that metrics mature either.